tactics/elimschemes.ml

```diff
@@ -27,39 +27,134 @@ let build_induction_scheme_in_type env dep sort ind =
   let sigma, pind = Evd.fresh_inductive_instance ~rigid:UState.univ_rigid env sigma ind in
   let sigma, sort = Evd.fresh_sort_in_quality ~rigid:UnivRigid sigma sort in
   let sigma, c = build_induction_scheme env sigma pind dep sort in
-  EConstr.to_constr sigma c, Evd.ustate sigma
+  Some (EConstr.to_constr sigma c, Evd.ustate sigma)
+
+let build_mutual_induction_scheme_in_type env dep sort isrec l =
+  let ind,_ = match l with | x::_ -> x | [] -> assert false in
+  let sigma, inst =
+    let _, ctx = Typeops.type_of_global_in_context env (Names.GlobRef.IndRef (ind,0)) in
+    let u, ctx = UnivGen.fresh_instance_from ctx None in
+    let u = EConstr.EInstance.make u in
+    let sigma = Evd.from_ctx (UState.of_context_set ctx) in
+    sigma, u
+  in
+  let n = List.length l in
+  let sigma, lrecspec =
+    let rec loop i n sigma ll =
+      if i>=n then (sigma,ll)
+      else
+        let new_sigma, new_sort = Evd.fresh_sort_in_quality ~rigid:UnivRigid sigma sort in
+        let (indd,ii) = List.nth l i in
+        let new_l = List.append ll [(((indd,ii),inst),dep,new_sort)] in
+        loop (i + 1) n new_sigma new_l
+    in
+    loop 0 n sigma []
+  in
+  let sigma, listdecl =
+    if isrec then Indrec.build_mutual_induction_scheme env sigma ~force_mutual:false lrecspec
+    else
+      List.fold_left_map (fun sigma (ind,dep,sort) ->
+          let sigma, c = Indrec.build_case_analysis_scheme env sigma ind dep sort in
+          let c, _ = Indrec.eval_case_analysis c in
+          sigma, c)
+        sigma lrecspec
+  in
+  let array = Array.of_list listdecl in
+  let l = Array.map (fun x -> EConstr.to_constr sigma x) array in
+  Some (l, Evd.ustate sigma)
+
+let make_suff_sort one_ind suff dep =
+  match one_ind with
+  | None -> suff
+  | Some i ->
+    let sort = i.mind_sort
+    in
+    match sort with
+    | Prop -> if dep then (Names.Id.to_string i.mind_typename) ^ "_" ^ suff ^ "_dep"
+      else (Names.Id.to_string i.mind_typename) ^ "_" ^ suff
+    | Type _ | SProp | Set -> if dep then (Names.Id.to_string i.mind_typename) ^ "_" ^ suff
+      else (Names.Id.to_string i.mind_typename) ^ "_" ^ suff ^ "_nodep"
+    | QSort _ -> (Names.Id.to_string i.mind_typename) ^ "_" ^ suff
 
 let rect_dep =
-  declare_individual_scheme_object "rect_dep"
-    (fun env _ x -> build_induction_scheme_in_type env true QualityOrSet.qtype x)
+  declare_individual_scheme_object (["Induction"], Some QualityOrSet.qtype)
+    (fun id -> make_suff_sort id "rect" true)
+    (fun env _ x _ -> build_induction_scheme_in_type env true QualityOrSet.qtype x)
+
+let mutual_rect_dep =
```

Comment:
If this is going to be rewritten, I am not sure if all these definitions should be kept.

Reply:
I don't understand why we define all this partial function.
Shouldn't indscheme just call `build_induction_scheme_in_type` directly ?
Unless tactics are still calling this functions on the fly rather than using `DeclareScheme`.